arch/xtensa/Kconfig

```diff
@@ -83,6 +83,7 @@ config ARCH_CHIP_ESP32S3
 	select ARCH_HAVE_RESET
 	select ARCH_HAVE_TESTSET
 	select ARCH_VECNOTIRQ
+	select LIBC_ARCH_FORCE_KERNEL if BUILD_PROTECTED
```

Comment:
@tmedicci but no LIBC_ARCH_xxx is selected if BUILD_PROTECTED equals y, so both kernel and userspace compile and link with the C version. Why LIBC_ARCH_FORCE_KERNEL is needed in this case?

Reply:
That is the point, @xiaoxiang781216 . If `BUILD_PROTECTED` equals `y`, `LIBC_ARCH_xxx` isn't selected and the software version would be built. But, in this case, we'd like to force the architecture-specific version to be used on kernel instead of the SW version (`LIBC_ARCH_FORCE_KERNEL` forces that). Alternatively, we could do that by changing the linker script, but that wouldn't be as clear for the user as creating this variable.

Reply:
so should you remove if BUILD_FLAT from select LIBC_ARCH_xxx in this patch too?

Reply:
After enabling memxxx chip version, we have three choice in PROTECTED or KERNEL mode:

1. Enable for both user and privilege mode
2. Enable for user mode only
3. Enable for privilege mode only

So, it's more suitable to have a choice with three options e.g.:
```
choice
	default LIBC_ARCH_BOTH

config LIBC_ARCH_BOTH

config LIBC_ARCH_KERNEL_ONLY

config LIBC_ARCH_USER_ONLY
```
And the check CONFIG_LIBC_ARCH_xxx in Make.defs need to move into source files, and Make.defs orsource code under libc/machine need to be updated too.

Reply:
No. Otherwise, the architecture-specific implementation would be used by the userspace.

Considering the protected build, the userspace is forbidden to access the architecture-specific (in this case, the FOM function). 

The considerations about this patch:

1) On flat build: there are no problems to use the architecture-specific implementation at all;
2) On protected build: we need to build the software version (that's why the `LIBC_ARCH_xxx` is not selected), but;
2.1) For the kernel, we can force the usage of the architecture-specific implementation by not building the software version when  `FORCE_LIBC_ARCH_KERNEL` equals `y` (the software version would be used only on the userspace).

Reply:
I'm sorry, I answered you before your last message (https://github.com/apache/nuttx/pull/9022#discussion_r1179518496). Let me think about it.

Reply:
> After enabling memxxx chip version, we have three choice in PROTECTED or KERNEL mode:
> 
> 1. Enable for both user and privilege mode
> 2. Enable for user mode only
> 3. Enable for privilege mode only
> 
> So, it's more suitable to have a choice with three options e.g.:
> 
> ```
> choice
> 	default LIBC_ARCH_BOTH
> 
> config LIBC_ARCH_BOTH
> 
> config LIBC_ARCH_KERNEL_ONLY
> 
> config LIBC_ARCH_USER_ONLY
> ```
> 
> And the check CONFIG_LIBC_ARCH_xxx in Make.defs need to move into source files, and Make.defs orsource code under libc/machine need to be updated too.

Well, I'd think something simpler (considering these variables you suggested).

First of all, please note that currently, there is no one-to-one relation between `CONFIG_LIBC_ARCH_xxx` and `libc/machine` source. Kconfig options from `libc/machine` select `CONFIG_LIBC_ARCH_xxx`, but the opposite is not true: selecting `CONFIG_LIBC_ARCH_xxx` won't select the `libc/machine` implementation. Instead, it'll only ensure that the software C version isn't being built. That is the case for ESP32-S3: we are not using the `XTENSA_xxx` version. We use the ROM functions whose symbols are provided by the linker script (`nuttx/boards/xtensa/esp32s3/common/scripts/esp32s3_rom.ld`): we only need to ensure that no other software has the same symbol.

That being said, currently 1) we cover the situation when the arch-specific version is not selected by providing an alternative software C version, but 2) the opposite, when arch-specific is selected, we can't ensure for sure that something will be built to provide these functions (we need either to provide symbols on linker or select the arch-specific implementation on `libc/machine`).

So, instead `CONFIG_FORCE_LIBC_ARCH_KERNEL` to ensure the software C version won't be built on kernel, we could use `CONFIG_LIBC_ARCH_BOTH/KERNEL_ONLY/USER_ONLY` to, in case of `CONFIG_LIBC_ARCH_xxx`  being selected, decide (depending whether the kernel is being built) to build or not the software C version. I wouldn't mess it with `libc/machine` sources/`Make.defs`, otherwise it'd be a huge modification and error-prone. My suggestion, in this case, affects only the same sources I already suggested modifying: just some different Karnaugh combinations. 

Honestly, It'd be just a better name specification than `CONFIG_FORCE_LIBC_ARCH_KERNEL`. But it's feasible.

What do you think about it? Am I missing something?

Reply:
Hi @xiaoxiang781216 , I've submitted an alternative implementation on https://github.com/apache/nuttx/pull/9138 (it's marked as a draft). 

Unfortunately, choices can't be selected directly by other configs, so I needed to provide supportive configs to enable the arch's chip to select it (I don't like the idea of leaving the decision of selecting a necessary option up to the user). Could you please evaluate it if it better suits the problem?